#! /bin/sh -e
## 008_src.dpatch by <unknown>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Various patches in src for the Debian package (before switch to dpatch)

if [ $# -lt 1 ]; then
    echo "`basename $0`: script expects -patch|-unpatch as argument" >&2
    exit 1
fi

[ -f debian/patches/00patch-opts ] && . debian/patches/00patch-opts
patch_opts="${patch_opts:--f --no-backup-if-mismatch} ${2:+-d $2}"

case "$1" in
    -patch) patch -p1 ${patch_opts} < $0;;
    -unpatch) patch -R -p1 ${patch_opts} < $0;;
    *)
        echo "`basename $0`: script expects -patch|-unpatch as argument" >&2
        exit 1;;
esac

exit 0

@DPATCH@

diff -Nru shadow-4.0.3/src/grpck.c shadow-4.0.3_30.4/src/grpck.c
--- shadow-4.0.3/src/grpck.c	2002-01-05 16:41:43.000000000 +0100
+++ shadow-4.0.3_30.4/src/grpck.c	2004-11-02 22:17:38.000000000 +0100
@@ -146,6 +146,7 @@
 	int errors = 0;
 	int deleted = 0;
 	int i;
+	int prune = 0;
 	struct commonio_entry *gre, *tgre;
 	struct group *grp;
 	int sort_mode = 0;
@@ -172,7 +173,7 @@
 	 * Parse the command line arguments
 	 */
 
-	while ((arg = getopt (argc, argv, "qrs")) != EOF) {
+	while ((arg = getopt (argc, argv, "qprs")) != EOF) {
 		switch (arg) {
 		case 'q':
 			/* quiet - ignored for now */
@@ -183,6 +184,9 @@
 		case 's':
 			sort_mode = 1;
 			break;
+		case 'p':
+			prune = 1;
+			break;
 		default:
 			usage ();
 		}
@@ -315,9 +319,13 @@
 			/*
 			 * prompt the user to delete the entry or not
 			 */
-
-			if (!yes_or_no ())
+			if (!prune) {
+			        if (!yes_or_no ())
+				        continue;
+			} else {
+			        puts("Yes");
 				continue;
+			}
 
 			/*
 			 * All group file deletions wind up here. This code
diff -Nru shadow-4.0.3/src/login.c shadow-4.0.3_30.4/src/login.c
--- shadow-4.0.3/src/login.c	2002-03-08 05:30:28.000000000 +0100
+++ shadow-4.0.3_30.4/src/login.c	2004-11-02 22:17:38.000000000 +0100
@@ -58,6 +58,11 @@
 #include <sys/proc.h>
 #include <sys/sysi86.h>
 #endif
+
+#ifndef MAXHOSTNAMELEN
+#define MAXHOSTNAMELEN 64
+#endif
+
 #ifdef RADIUS
 /*
  * Support for RADIUS authentication based on a hacked util-linux login
@@ -289,6 +294,8 @@
 	for (arg = 1; arg < argc; arg++) {
 		if (argv[arg][0] == '-' && strlen (argv[arg]) > 2)
 			usage ();
+		if (!strcmp(argv[arg], "--"))
+			break; /* stop checking on a "--" */
 	}
 }
 
@@ -466,7 +473,7 @@
 
 	check_flags (argc, argv);
 
-	while ((flag = getopt (argc, argv, "d:f:h:pr:")) != EOF) {
+	while ((flag = getopt (argc, argv, "d:f::h:pr:")) != EOF) {
 		switch (flag) {
 		case 'p':
 			pflg++;
@@ -475,11 +482,16 @@
 			/*
 			 * username must be a separate token
 			 * (-f root, *not* -froot).  --marekm
+ 			 *
+ 			 * if -f has an arg, use that, else use the
+ 			 * normal user name passed after all options
+ 			 * --benc
 			 */
-			if (optarg != argv[optind - 1])
+			if (optarg != NULL && optarg != argv[optind - 1])
 				usage ();
 			fflg++;
-			STRFCPY (username, optarg);
+			if (optarg)
+			  STRFCPY (username, optarg);
 			break;
 #ifdef	RLOGIN
 		case 'r':
@@ -633,7 +645,7 @@
 		init_env ();
 
 		if (optind < argc) {	/* get the user name */
-			if (rflg || fflg)
+			if (rflg || (fflg && username[0]))
 				usage ();
 
 #ifdef SVR4
@@ -763,49 +775,98 @@
 			 * MAX_LOGIN_TRIES?
 			 */
 
-			retcode = pam_authenticate (pamh, 0);
-			while ((failcount++ < retries) &&
-			       ((retcode == PAM_AUTH_ERR) ||
-				(retcode == PAM_USER_UNKNOWN) ||
-				(retcode == PAM_CRED_INSUFFICIENT) ||
-				(retcode == PAM_AUTHINFO_UNAVAIL))) {
-				pam_get_item (pamh, PAM_USER,
-					      (const void **) &pam_user);
-				syslog (LOG_NOTICE,
-					"FAILED LOGIN %d FROM %s FOR %s, %s",
-					failcount, hostname, pam_user,
-					pam_strerror (pamh, retcode));
-#ifdef HAVE_PAM_FAIL_DELAY
-				pam_fail_delay (pamh, 1000000 * delay);
+			failcount = 0;
+			while (1) {
+			  const char *failent_user;
+			  failed = 0;
+			  
+			  failcount++;
+			  if (delay > 0)
+			    retcode = pam_fail_delay(pamh, 1000000*delay);
+			  
+			  retcode = pam_authenticate (pamh, 0);
+			  
+			  pam_get_item (pamh, PAM_USER,
+					(const void **) &pam_user);
+			  
+			  if (pam_user && pam_user[0]) {
+			    pwd = getpwnam(pam_user);
+			    if (pwd) {
+			      pwent = *pwd;
+			      failent_user = pwent.pw_name;
+			    } else {
+			      if (getdef_bool("LOG_UNKFAIL_ENAB") && pam_user)
+				failent_user = pam_user;
+			      else
+				failent_user = "UNKNOWN";
+			    }
+			  } else {
+			    pwd = NULL;
+			    failent_user = "UNKNOWN";
+			  }
+			  
+			  if (retcode == PAM_MAXTRIES || failcount >= retries) {
+			    syslog (LOG_NOTICE,
+				    _("TOO MANY LOGIN TRIES (%d)%s FOR `%s'"),
+				    failcount, fromhost, failent_user);
+#ifndef USE_PAM
+			    if (pwd && getdef_bool("FAILLOG_ENAB"))
+			      failure (pwent.pw_uid, tty, &faillog);
+#endif
+			    fprintf(stderr,
+				    _("Maximum number of tries exceeded (%d)\n"),
+				    failcount);
+			    PAM_END;
+			    exit(0);
+			  } else if (retcode == PAM_ABORT) {
+			    /* Serious problems, quit now */
+			    fprintf(stderr,_("login: abort requested by PAM\n"));
+			    syslog(LOG_ERR,_("PAM_ABORT returned from pam_authenticate()"));
+			    PAM_END;
+			    exit(99);
+			  } else if (retcode != PAM_SUCCESS) {
+			    syslog(LOG_NOTICE,_("FAILED LOGIN (%d)%s FOR `%s', %s"),
+				   failcount, fromhost, failent_user,
+				   pam_strerror (pamh, retcode));
+			    failed = 1;
+			  }
+#ifndef USE_PAM
+			  if (pwd && getdef_bool("FAILLOG_ENAB") &&
+			      ! failcheck (pwent.pw_uid, &faillog, failed)) {
+			    SYSLOG((LOG_CRIT, FAILURE_CNT, failent_user, fromhost));
+			    failed = 1;
+			  }
 #endif
-				fprintf (stderr, "Login incorrect\n\n");
-				pam_set_item (pamh, PAM_USER, NULL);
-				retcode = pam_authenticate (pamh, 0);
-			}
 
-			if (retcode != PAM_SUCCESS) {
-				pam_get_item (pamh, PAM_USER,
-					      (const void **) &pam_user);
-
-				if (retcode == PAM_MAXTRIES)
-					syslog (LOG_NOTICE,
-						"TOO MANY LOGIN TRIES (%d) FROM %s FOR %s, %s",
-						failcount, hostname,
-						pam_user,
-						pam_strerror (pamh,
-							      retcode));
-				else
-					syslog (LOG_NOTICE,
-						"FAILED LOGIN SESSION FROM %s FOR %s, %s",
-						hostname, pam_user,
-						pam_strerror (pamh,
-							      retcode));
-
-				fprintf (stderr, "\nLogin incorrect\n");
-				pam_end (pamh, retcode);
-				exit (0);
+			  if (!failed)
+			    break;
+			  
+			  fprintf(stderr,"Login incorrect\n\n");
+#ifndef USE_PAM
+			  if (pwd && getdef_bool("FAILLOG_ENAB"))
+			    failure (pwent.pw_uid, tty, &faillog);
+#endif
+			  if (getdef_str("FTMP_FILE") != NULL) {
+#if HAVE_UTMPX_H
+			    failent = utxent;
+			    gettimeofday(&(failent.ut_tv), NULL);
+#else
+			    failent = utent;
+			    time(&failent.ut_time);
+#endif
+			    strncpy(failent.ut_user, failent_user, sizeof(failent.ut_user));
+#ifdef USER_PROCESS
+			    failent.ut_type = USER_PROCESS;
+#endif
+			    failtmp(&failent);
+			  }
+			  
+			  /* Let's give it another go around */
+			  pam_set_item(pamh,PAM_USER,NULL);
 			}
 
+			/* We don't get here unless they were authenticated above */
+			alarm(0);
 			retcode = pam_acct_mgmt (pamh, 0);
 
 			if (retcode == PAM_NEW_AUTHTOK_REQD) {
@@ -828,11 +889,14 @@
 
 		if (!pwd || setup_groups (pwd))
 			exit (1);
+		else
+		        pwent = *pwd;
 
 		retcode = pam_setcred (pamh, PAM_ESTABLISH_CRED);
 		PAM_FAIL_CHECK;
 
-		retcode = pam_open_session (pamh, 0);
+		retcode = pam_open_session (pamh,
+					    hushed(&pwent) ? PAM_SILENT : 0);
 		PAM_FAIL_CHECK;
 
 #else				/* ! USE_PAM */
@@ -1002,6 +1066,7 @@
 			failed = 1;
 		}
 #endif
+#ifndef USE_PAM
 		if (pwd && getdef_bool ("FAILLOG_ENAB") &&
 		    !failcheck (pwent.pw_uid, &faillog, failed)) {
 			SYSLOG ((LOG_CRIT, 
@@ -1009,12 +1074,15 @@
 				 username, fromhost));
 			failed = 1;
 		}
+#endif
 		if (!failed)
 			break;
 
+#ifndef USE_PAM
 		/* don't log non-existent users */
 		if (pwd && getdef_bool ("FAILLOG_ENAB"))
 			failure (pwent.pw_uid, tty, &faillog);
+#endif
 		if (getdef_str ("FTMP_FILE") != NULL) {
 			const char *failent_user;
 
@@ -1208,6 +1276,40 @@
 	login_fbtab (tty, pwent.pw_uid, pwent.pw_gid);
 #endif
 
+#ifdef USE_PAM
+	/*
+	 * We must fork before setuid() because we need to call
+	 * pam_close_session() as root.
+	 *
+	 * Note: not true in other (non-Linux) PAM implementations, where
+	 * the parent process of login (init, telnetd, ...) is responsible
+	 * for calling pam_close_session(). This avoids an extra process for
+	 * each login. Maybe we should do this on Linux too? We let the
+	 * admin configure whether they need to keep login around to close
+	 * sessions.
+	 */
+	if (getdef_bool ("CLOSE_SESSIONS")) {
+		signal (SIGINT, SIG_IGN);
+		child = fork ();
+		if (child < 0) {
+			/* error in fork() */
+			fprintf (stderr,
+				 "login: failure forking: %s",
+				 strerror (errno));
+			PAM_END;
+			exit (0);
+		} else if (child) {
+			/*
+			 * parent - wait for child to finish, then cleanup
+			 * session
+			 */
+			wait (NULL);
+			PAM_END;
+			exit (0);
+		}
+		/* child */
+	}
+#endif
 	/* We call set_groups() above because this clobbers pam_groups.so */
 #ifndef USE_PAM
 	if (setup_uid_gid (&pwent, is_console))
@@ -1309,41 +1410,6 @@
 	signal (SIGTERM, SIG_DFL);	/* default terminate signal */
 	signal (SIGALRM, SIG_DFL);	/* default alarm signal */
 	signal (SIGHUP, SIG_DFL);	/* added this.  --marekm */
-
-#ifdef USE_PAM
-	/*
-	 * We must fork before setuid() because we need to call
-	 * pam_close_session() as root.
-	 *
-	 * Note: not true in other (non-Linux) PAM implementations, where
-	 * the parent process of login (init, telnetd, ...) is responsible
-	 * for calling pam_close_session(). This avoids an extra process for
-	 * each login. Maybe we should do this on Linux too? We let the
-	 * admin configure whether they need to keep login around to close
-	 * sessions.
-	 */
-	if (getdef_bool ("CLOSE_SESSIONS")) {
-		signal (SIGINT, SIG_IGN);
-		child = fork ();
-		if (child < 0) {
-			/* error in fork() */
-			fprintf (stderr,
-				 "login: failure forking: %s",
-				 strerror (errno));
-			PAM_END;
-			exit (0);
-		} else if (child) {
-			/*
-			 * parent - wait for child to finish, then cleanup
-			 * session
-			 */
-			wait (NULL);
-			PAM_END;
-			exit (0);
-		}
-		/* child */
-	}
-#endif
 	signal (SIGINT, SIG_DFL);	/* default interrupt signal */
 
 	endpwent ();	/* stop access to password file */
@@ -1357,7 +1440,11 @@
 	if (pwent.pw_uid == 0)
 		SYSLOG ((LOG_NOTICE, "ROOT LOGIN %s", fromhost));
 	else if (getdef_bool ("LOG_OK_LOGINS"))
+#ifdef USE_PAM
+		SYSLOG ((LOG_INFO, "`%s' logged in %s", pam_user, fromhost));
+#else
 		SYSLOG ((LOG_INFO, "`%s' logged in %s", username, fromhost));
+#endif
 	closelog ();
 #ifdef RADIUS
 	if (is_rad_login) {
diff -Nru shadow-4.0.3/src/newgrp.c shadow-4.0.3_30.4/src/newgrp.c
--- shadow-4.0.3/src/newgrp.c	2002-01-06 16:00:07.000000000 +0100
+++ shadow-4.0.3_30.4/src/newgrp.c	2004-11-02 22:17:38.000000000 +0100
@@ -91,7 +91,7 @@
 
 #if ENABLE_NLS
 	/* XXX - remove when gettext is safe to use in setuid programs */
-	sanitize_env ();
+	/* sanitize_env ();*/
 #endif
 
 	setlocale (LC_ALL, "");
@@ -386,8 +386,13 @@
 		SYSLOG ((LOG_INFO, "user `%s' switched to group `%s'",
 			 name, group));
 	if (getdef_bool ("SYSLOG_SG_ENAB")) {
-		char *loginname = xstrdup (getlogin ());
-		char *tty = xstrdup (ttyname (0));
+		char *loginname = getlogin ();
+		char *tty = ttyname (0);
+
+		if (loginname != NULL)
+		  loginname = xstrdup (loginname);
+		if (tty != NULL)
+		  tty = xstrdup (tty);
 
 		if (loginname == NULL)
 			loginname = "???";
diff -Nru shadow-4.0.3/src/su.c shadow-4.0.3_30.4/src/su.c
--- shadow-4.0.3/src/su.c	2002-03-08 05:30:28.000000000 +0100
+++ shadow-4.0.3_30.4/src/su.c	2004-11-02 22:17:38.000000000 +0100
@@ -26,6 +26,24 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  */
+/* Some parts substantially derived from an ancestor of: */
+/* su for GNU.  Run a shell with substitute user and group IDs.
+   Copyright (C) 1992-2003 Free Software Foundation, Inc.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2, or (at your option)
+   any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software Foundation,
+   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */
+
 
 #include <config.h>
 
@@ -49,6 +67,7 @@
 #include <grp.h>
 #include <signal.h>
 #include <pwd.h>
+#include <getopt.h>
 #include "pwauth.h"
 #include "getdef.h"
 
@@ -79,6 +98,18 @@
 
 /* local function prototypes */
 
+/* If nonzero, change some environment vars to indicate the user su'd to.  */
+static int change_environment;
+
+static struct option const longopts[] =
+{
+  {"command", required_argument, 0, 'c'},
+  {"preserve-environment", no_argument, 0, 'p'},
+  {"shell", required_argument, 0, 's'},
+  {"help", no_argument, 0, 'h'},
+  {0, 0, 0, 0}
+};
+
 #ifndef USE_PAM
 
 static RETSIGTYPE die (int);
@@ -118,6 +149,96 @@
 }
 #endif				/* !USE_PAM */
 
+/* borrowed from GNU sh-utils' "su.c" */
+static int
+restricted_shell (const char *shell)
+{
+       char *line;
+
+       setusershell ();
+       while ((line = getusershell ()) != NULL) {
+               if (*line != '#' && strcmp (line, shell) == 0) {
+                       endusershell ();
+                       return 0;
+               }
+       }
+       endusershell ();
+       return 1;
+}
+
+/* borrowed from GNU sh-utils' "su.c" */
+static int
+elements (char **arr)
+{
+  int n = 0;
+
+  for (n = 0; *arr; ++arr)
+    ++n;
+  return n;
+}
+
+/* borrowed from GNU sh-utils' "su.c" */
+static void
+run_shell (char *shell, const char *command, char **additional_args, int login)
+{
+  const char **args;
+  int argno = 1;
+  char cmd[BUFSIZ];
+  int cmd_len_left = sizeof(cmd) - 1;
+
+  cmd[0] = '\0';
+
+  if (additional_args)
+    args = (const char **) xmalloc (sizeof (char *)
+                                    * (10 + elements (additional_args)));
+  else
+    args = (const char **) xmalloc (sizeof (char *) * 10);
+
+  if (login)
+    {
+      char *arg0;
+      char *shell_basename;
+
+      shell_basename = getdef_str("SU_NAME");
+      if (!shell_basename)
+       shell_basename = Basename(shell);
+
+      arg0 = xmalloc (strlen (shell_basename) + 2);
+      arg0[0] = '-';
+      strcpy (arg0 + 1, shell_basename);
+      args[0] = arg0;
+    }
+  else
+    args[0] = Basename(shell);
+  if (command || additional_args)
+    args[argno++] = "-c";
+  if (command) {
+    if (strlen(command) > cmd_len_left) {
+      fprintf(stderr, _("Command line args too long\n"));
+      exit(1);
+    }
+    strcat(cmd, command);
+    cmd_len_left -= strlen(command);
+  }
+  if (additional_args)
+    for (; *additional_args; ++additional_args) {
+      if ((strlen(*additional_args) + 1) > cmd_len_left) {
+	fprintf(stderr, _("Command line args too long\n"));
+	exit(1);
+      }
+      if (cmd[0]) {
+	strcat(cmd, " ");
+	cmd_len_left--;
+      }
+      strcat(cmd, *additional_args);
+      cmd_len_left -= strlen(*additional_args);
+    }
+  if (cmd[0]) args[argno++] = cmd;
+  args[argno] = NULL;
+  execv (shell, (char **) args);
+  fprintf (stderr, _("No shell\n"));
+  SYSLOG((LOG_WARN, "Cannot execute %s\n", shell));
+}
 
 static void su_failure (const char *tty)
 {
@@ -125,7 +246,7 @@
 #ifdef USE_SYSLOG
 	if (getdef_bool ("SYSLOG_SU_ENAB"))
 		SYSLOG ((pwent.pw_uid ? LOG_INFO : LOG_NOTICE,
-			 "- %s %s-%s", tty,
+			 "- %s %s:%s", tty,
 			 oldname[0] ? oldname : "???",
 			 name[0] ? name : "???"));
 	closelog ();
@@ -153,13 +274,14 @@
 {
 	char *cp;
 	const char *tty = 0;	/* Name of tty SU is run from        */
-	int doshell = 0;
 	int fakelogin = 0;
 	int amroot = 0;
 	uid_t my_uid;
 	struct passwd *pw = 0;
 	char **envp = environ;
-
+	char *command = 0, *shell = 0, **additional_args = 0;
+	int optc;
+	char *tmp_name;
 #ifdef USE_PAM
 	int ret;
 #else				/* !USE_PAM */
@@ -174,12 +296,14 @@
 #endif
 #endif				/* !USE_PAM */
 
-	sanitize_env ();
+	/* sanitize_env (); */
 
 	setlocale (LC_ALL, "");
 	bindtextdomain (PACKAGE, LOCALEDIR);
 	textdomain (PACKAGE);
 
+	change_environment = 1;
+
 	/*
 	 * Get the program name. The program name is used as a prefix to
 	 * most error messages.
@@ -224,15 +348,67 @@
 	 * Process the command line arguments. 
 	 */
 
-	argc--;
-	argv++;			/* shift out command name */
-
-	if (argc > 0 && strcmp (argv[0], "-") == 0) {
+	while ((optc = getopt_long (argc, argv, "c:mps:h", longopts, NULL)) != -1) {
+		switch (optc) {
+		    case 0:
+			break;
+		    case 'c':
+			command = optarg;
+			break;
+		    case 'm':
+		    case 'p':
+			change_environment = 0;
+			break;
+		    case 's':
+			shell = optarg;
+			break;
+		    default:
+			fprintf(stderr, _("\
+Usage: su [OPTS] [-] [username [ARGS]]\n\
+	-	make this a login shell\n\
+	-c, --command=<command>\n\
+		pass command to the invoked shell using its -c\n\
+		option\n\
+       -m, -p, --preserve-environment\n\
+		do not reset environment variables, and keep the\n\
+		same shell\n\
+       -s, --shell=<shell>\n\
+		use shell instead of the default in /etc/passwd\n"));
+			exit(1);
+		}
+	}
+	
+        if (optind < argc && !strcmp (argv[optind], "-")) {
 		fakelogin = 1;
-		argc--;
-		argv++;		/* shift ... */
+		++optind;
 	}
 
+	if (optind < argc)
+		strncpy(name, argv[optind++], sizeof(name) - 1);
+	else {
+	        struct passwd *root_pw = getpwuid(0);
+		if (root_pw == NULL) {
+		  SYSLOG((LOG_CRIT, "There is no UID 0 user."));
+		  su_failure(tty);
+		}
+                strcpy(name, root_pw->pw_name);
+	}
+
+	if (optind < argc)
+		additional_args = argv + optind;
+
+	/*
+	 * Get the user's real name.  The current UID is used to determine
+	 * who has executed su.  That user ID must exist.
+	 */
+
+	pw = get_my_pwent();
+	if (!pw) {
+		SYSLOG((LOG_CRIT, "Unknown UID: %d\n", (int) my_uid));
+		su_failure(tty);
+	}
+	STRFCPY(oldname, pw->pw_name);
+
 	/*
 	 * If a new login is being set up, the old environment will be
 	 * ignored and a new one created later on.
@@ -257,35 +433,6 @@
 			addenv (*envp++, NULL);
 	}
 
-	/*
-	 * The next argument must be either a user ID, or some flag to a
-	 * subshell. Pretty sticky since you can't have an argument which
-	 * doesn't start with a "-" unless you specify the new user name.
-	 * Any remaining arguments will be passed to the user's login shell.
-	 */
-
-	if (argc > 0 && argv[0][0] != '-') {
-		STRFCPY (name, argv[0]);	/* use this login id */
-		argc--;
-		argv++;		/* shift ... */
-	}
-	if (!name[0])		/* use default user ID */
-		(void) strcpy (name, "root");
-
-	doshell = argc == 0;	/* any arguments remaining? */
-
-	/*
-	 * Get the user's real name. The current UID is used to determine
-	 * who has executed su. That user ID must exist.
-	 */
-
-	pw = get_my_pwent ();
-	if (!pw) {
-		SYSLOG ((LOG_CRIT, "Unknown UID: %u", my_uid));
-		su_failure (tty);
-	}
-	STRFCPY (oldname, pw->pw_name);
-
 #ifndef USE_PAM
 #ifdef SU_ACCESS
 	/*
@@ -399,9 +546,17 @@
 	 * Set the default shell.
 	 */
 
-	if (pwent.pw_shell[0] == '\0')
-		pwent.pw_shell = "/bin/sh";	/* XXX warning: const */
+	if (pwent.pw_shell == NULL || pwent.pw_shell[0] == '\0')
+		pwent.pw_shell = (char *) "/bin/sh";
+
+	if (shell == 0 && change_environment == 0)
+		shell = getenv ("SHELL");
+	if (shell != 0 && getuid () && restricted_shell (pwent.pw_shell))
+		shell = 0;
+	if (shell == 0)
+	shell = (char *) strdup (pwent.pw_shell);
 
+	signal(SIGINT, SIG_IGN);
 #ifdef USE_PAM
 	ret = pam_authenticate (pamh, 0);
 	if (ret != PAM_SUCCESS) {
@@ -427,6 +582,14 @@
 			su_failure (tty);
 		}
 	}
+	ret = pam_get_item(pamh, PAM_USER, (const void **) &tmp_name);
+	if (ret != PAM_SUCCESS) {
+		SYSLOG((LOG_ERR, "pam_get_item: internal PAM error\n"));
+		fprintf(stderr, "%s: Internal PAM error retrieving username\n", Prog);
+		pam_end(pamh, ret);
+		su_failure(tty);
+	}
+	strncpy(name, tmp_name, sizeof(name) - 1);
 #else				/* !USE_PAM */
 	/*
 	 * Set up a signal handler in case the user types QUIT.
@@ -507,10 +670,14 @@
 	}
 #endif
 
-	environ = newenvp;	/* make new environment active */
-
-	if (getenv ("IFS"))	/* don't export user IFS ... */
-		addenv ("IFS= \t\n", NULL);	/* ... instead, set a safe IFS */
+	if (change_environment || restricted_shell(pwent.pw_shell)) {
+		environ = newenvp;			/* make new environment active */
+		if (getenv ("IFS"))			/* don't export user IFS ... */
+			addenv("IFS= \t\n", NULL);	/* ... instead, set a safe IFS */
+	} else {
+		if (getenv ("IFS"))
+			putenv("IFS= \t\n");
+	}
 
 	if (pwent.pw_shell[0] == '*') {	/* subsystem root required */
 		pwent.pw_shell++;	/* skip the '*' */
@@ -529,7 +696,7 @@
 #endif
 #ifdef USE_SYSLOG
 	if (getdef_bool ("SYSLOG_SU_ENAB"))
-		SYSLOG ((LOG_INFO, "+ %s %s-%s", tty,
+		SYSLOG ((LOG_INFO, "+ %s %s:%s", tty,
 			 oldname[0] ? oldname : "???",
 			 name[0] ? name : "???"));
 #endif
@@ -554,17 +721,56 @@
 		pam_end (pamh, ret);
 		exit (1);
 	}
+	ret = pam_open_session(pamh, 0);
+	if (ret != PAM_SUCCESS) {
+		SYSLOG((LOG_ERR, "pam_open_session: %s\n", pam_strerror(pamh, ret)));
+		fprintf(stderr, "%s: %s\n", Prog, pam_strerror(pamh, ret));
+		pam_setcred(pamh, PAM_DELETE_CRED);
+		pam_end(pamh, ret);
+		exit(1);
+	}
+	/* We must fork before setuid() because we need to call
+	 * pam_close_session() as root.
+	 */
+
+	/* We let the admin configure whether they need to keep login
+	   around to close sessions */
+	if (getdef_bool("CLOSE_SESSIONS")) {
+		pid_t pid;
+		int status;
+
+		signal(SIGINT, SIG_IGN);
+		pid = fork();
+
+		switch(pid) {
+		case -1:
+			SYSLOG((LOG_ERR, "su: fork failure: %m"));
+			perror("su: fork failure");
+			pam_setcred(pamh, PAM_DELETE_CRED);
+			pam_close_session(pamh, 0);
+			pam_end(pamh, PAM_ABORT);
+			exit(1);
+		case 0: /* child */
+			signal(SIGINT, SIG_DFL);
+			break;
+		default: /* parent */
+			waitpid(pid, &status, 0);
+			/* now we are done using PAM */
+			pam_setcred(pamh, PAM_DELETE_CRED);
+			ret = pam_close_session(pamh, 0);
+			pam_end(pamh, ret);
+			exit(WEXITSTATUS(status));
+		}
+	}
 
 	/* become the new user */
 	if (change_uid (&pwent)) {
+		pam_close_session(pamh, 0);
 		pam_setcred (pamh, PAM_DELETE_CRED);
 		pam_end (pamh, PAM_ABORT);
 		exit (1);
 	}
 
-	/* now we are done using PAM */
-	pam_end (pamh, PAM_SUCCESS);
-
 #else				/* !USE_PAM */
 	if (!amroot)		/* no limits if su from root */
 		setup_limits (&pwent);
@@ -573,11 +779,14 @@
 		exit (1);
 #endif				/* !USE_PAM */
 
-	if (fakelogin)
-		setup_env (&pwent);
+	if (fakelogin && (change_environment || restricted_shell(pwent.pw_shell)))
+		setup_env(&pwent);
 #if 1				/* Suggested by Joey Hess. XXX - is this right?  */
-	else
-		addenv ("HOME", pwent.pw_dir);
+	else if (change_environment || restricted_shell(pwent.pw_shell)) {
+		addenv("HOME", pwent.pw_dir);
+		addenv("USER", pwent.pw_name);
+		addenv("SHELL", shell);
+	}
 #endif
 
 	/*
@@ -589,46 +798,6 @@
 	 */
 	closelog ();
 
-	/*
-	 * See if the user has extra arguments on the command line. In that
-	 * case they will be provided to the new user's shell as arguments.
-	 */
-
-	if (fakelogin) {
-		char *arg0;
-
-#if 0				/* XXX - GNU su doesn't do this.  --marekm */
-		if (!hushed (&pwent)) {
-			motd ();
-			mailcheck ();
-		}
-#endif
-		cp = getdef_str ("SU_NAME");
-		if (!cp)
-			cp = Basename (pwent.pw_shell);
-
-		arg0 = xmalloc (strlen (cp) + 2);
-		arg0[0] = '-';
-		strcpy (arg0 + 1, cp);
-		cp = arg0;
-	} else
-		cp = Basename (pwent.pw_shell);
-
-	if (!doshell) {
-
-		/*
-		 * Use new user's shell from /etc/passwd and create an argv
-		 * with the rest of the command line included.
-		 */
-
-		argv[-1] = pwent.pw_shell;
-		(void) execv (pwent.pw_shell, &argv[-1]);
-		(void) fprintf (stderr, _("No shell\n"));
-		SYSLOG ((LOG_WARN, "Cannot execute %s", pwent.pw_shell));
-		closelog ();
-		exit (1);
-	}
-
-	shell (pwent.pw_shell, cp);
+	run_shell (shell, command, additional_args, fakelogin);
 	 /*NOTREACHED*/ exit (1);
 }
diff -Nru shadow-4.0.3/src/userdel.c shadow-4.0.3_30.4/src/userdel.c
--- shadow-4.0.3/src/userdel.c	2002-01-05 16:41:44.000000000 +0100
+++ shadow-4.0.3_30.4/src/userdel.c	2004-11-02 22:17:38.000000000 +0100
@@ -147,6 +147,7 @@
 	struct group *ngrp;
 
 #ifdef	SHADOWGRP
+	int deleted_user_group = 0;
 	const struct sgrp *sgrp;
 	struct sgrp *nsgrp;
 #endif				/* SHADOWGRP */
@@ -209,6 +210,10 @@
 
 		gr_remove (grp->gr_name);
 
+#ifdef SHADOWGRP
+		deleted_user_group = 1;
+#endif
+
 		/*
 		 * Update the DBM group file with the new entry as well.
 		 */
@@ -279,6 +284,10 @@
 		SYSLOG ((LOG_INFO, "delete `%s' from shadow group `%s'\n",
 			 user_name, nsgrp->sg_name));
 	}
+
+	if (deleted_user_group)
+		sgr_remove(user_name);
+
 #ifdef	NDBM
 	endsgent ();
 #endif				/* NDBM */
diff -Nru shadow-4.0.3/src/usermod.c shadow-4.0.3_30.4/src/usermod.c
--- shadow-4.0.3/src/usermod.c	2002-01-05 16:41:44.000000000 +0100
+++ shadow-4.0.3_30.4/src/usermod.c	2004-11-02 22:17:38.000000000 +0100
@@ -1544,9 +1544,14 @@
 				if (copy_tree (user_home, user_newhome,
 					       uflg ? user_newid : -1,
 					       gflg ? user_newgid : -1) ==
-				    0 && remove_tree (user_home) == 0
-				    && rmdir (user_home) == 0)
-					return;
+				    0) {
+				  if (remove_tree (user_home) != 0 ||
+				    rmdir (user_home) != 0)
+				    fprintf (stderr,
+					     _("%s: warning: failed to completely remove old home directory %s"),
+					     Prog, user_home);
+				  return;
+				}
 
 				(void) remove_tree (user_newhome);
 				(void) rmdir (user_newhome);
